grass/kernel.c

```diff
@@ -108,78 +113,68 @@ static void proc_yield() {
     proc_set_running(curr_pid);
 }
 
-static void proc_send(struct syscall *sc) {
-    sc->msg.sender = curr_pid;
-    int receiver = sc->msg.receiver;
-
-    for (uint i = 0; i < MAX_NPROCESS; i++)
-        if (proc_set[i].pid == receiver) {
-            /* Find the receiver */
-            if (proc_set[i].status != PROC_WAIT_TO_RECV) {
-                curr_status = PROC_WAIT_TO_SEND;
-                proc_set[proc_curr_idx].receiver_pid = receiver;
-            } else {
-                /* Copy message from sender to kernel stack */
-                struct sys_msg tmp;
-                earth->mmu_switch(curr_pid);
-                memcpy(&tmp, &sc->msg, sizeof(tmp));
-
-                /* Copy message from kernel stack to receiver */
-                earth->mmu_switch(receiver);
-                memcpy(&sc->msg, &tmp, sizeof(tmp));
-
-                /* Set receiver process as runnable */
-                proc_set_runnable(receiver);
-            }
-            proc_yield();
-            return;
+static int proc_send(struct syscall *sc, struct process *sender) {
+    if (pending_ipc_buffer->in_use == 1) return -1;
+    
+    for (uint i = 0; i < MAX_NPROCESS; i++) {
+        struct process dst = proc_set[i];
+        if (dst.pid == sc->msg.receiver) {
+            /* Destination is not receiving, or will not take msg from sender */
+            if (! (dst.status == PROC_PENDING && dst.pending_syscall == PENDING_RECV) ) return -1;
+            if (! (dst.receive_from == GPID_ALL || dst.receive_from == sender->pid) ) return -1;
+            
+            pending_ipc_buffer->in_use = 1;
+            pending_ipc_buffer->sender = sender->pid;
+            pending_ipc_buffer->receiver = sc->msg.receiver;
+
+            memcpy(pending_ipc_buffer->msg, sc->msg.content, sizeof(sc->msg.content));
+            return 0;
         }
+    }
 
-    sc->retval = -1;
+    return -2; // Error, receiver does not exist
```

Comment:
I think `return -2` here is ugly. The issue here is how to deal with `sc->retval`. I feel like we can simply (1) remove `retval` from `struct syscall`; (2) make `sys_send` and `sys_recv` return void just like `sys_exit`. Then, we can simply return 0 here. We are not really using `retval` anyways.

Reply:
In the first commit, `proc_try_send`, `sys_send`, and `sys_recv` were changed to return void and print FATAL messages instead of returning error codes.

The third commit removes them for memory in the grass layer. 

Do you believe printing is necessary to properly catch errors?